awx_collection/plugins/module_utils/controller_api.py

```diff
@@ -700,17 +678,20 @@ def modify_associations(self, association_endpoint, new_association_list):
         response = self.get_all_endpoint(association_endpoint)
         existing_associated_ids = [association['id'] for association in response['json']['results']]
 
-        # Disassociate anything that is in existing_associated_ids but not in new_association_list
-        ids_to_remove = list(set(existing_associated_ids) - set(new_association_list))
-        for an_id in ids_to_remove:
+        # If the current associations match the desired associations then we can just return
+        if existing_associated_ids == new_association_list:
+            return
+
+        # Some associations can be ordered (like galaxy credentials), because of this we have to remove everything to re-add in order
+        for an_id in existing_associated_ids:
             response = self.post_endpoint(association_endpoint, **{'data': {'id': int(an_id), 'disassociate': True}})
             if response['status_code'] == 204:
                 self.json_output['changed'] = True
             else:
                 self.fail_json(msg="Failed to disassociate item {0}".format(response['json'].get('detail', response['json'])))
 
-        # Associate anything that is in new_association_list but not in `association`
-        for an_id in list(set(new_association_list) - set(existing_associated_ids)):
+        # Now we will add everything back in the received order
```

Comment:
I understand that this is what the UI does for ordered relationships (like galaxy credentials), and for those types of relationships, I agree, this should also be acceptable in the API.

But if we ran this for _unordered_ relationships, that would be significantly disruptive to users. Because the order we get from the API isn't deterministic, it will remove all objects and add them back arbitrarily. This will dramatically slow down playbooks, change task status from ok to changed incorrectly, and spam the activity stream with pointless removals and associations.

...if I'm understand this.

I don't see a choice other than distinguishing ordered from unordered relationships.

Reply:
Yea, we'd have to introduce a new field to indicate what is ordered and not right? There is nothing in the API which would tell us now?

Reply:
yeah that's right. I can tell you the names of ordered relationships - it's "instance_groups" and "input_inventories"

Reply:
Added https://github.com/ansible/awx/issues/13654 with a back reference here.